terraform/modules/PersistGrantsGovXMLDB/main.tf

```
@@ -0,0 +1,117 @@
+  )
+}
+
+data "aws_s3_bucket" "prepared_data" {
+  bucket = var.grants_prepared_data_bucket_name
+}
+
+module "lambda_execution_policy" {
+  source  = "cloudposse/iam-policy/aws"
+  version = "0.4.0"
+
+  iam_source_policy_documents = var.additional_lambda_execution_policy_documents
+  iam_policy_statements = {
+    AllowInspectS3PreparedData = {
+      effect = "Allow"
+      actions = [
+        "s3:GetObject",
+        "s3:ListBucket"
+      ]
+      resources = [
+        data.aws_s3_bucket.prepared_data.arn,
+        "${data.aws_s3_bucket.prepared_data.arn}/*/*/grants.gov/v2.xml"
+      ]
+    }
+    AllowDynamoDBPreparedData = {
+      effect = "Allow"
+      actions = [
+        "dynamodb:ListTables",
+        "dynamodb:BatchGetItem",
+        "dynamodb:GetItem",
+        "dynamodb:Query",
+        "dynamodb:Scan",
+        "dynamodb:BatchWriteItem",
+        "dynamodb:PutItem",
+        "dynamodb:UpdateItem"
+      ]
+      resources = [var.grants_prepared_dynamodb_table_arn]
+    }
+  }
+}
+
+resource "aws_s3_bucket_notification" "default" {
+  bucket = data.aws_s3_bucket.prepared_data.id
+
+  lambda_function {
+    lambda_function_arn = module.lambda_function.lambda_function_arn
+    events              = ["s3:ObjectCreated:*"]
+    filter_suffix       = "/grants.gov/v2.xml"
+  }
+}
+
+module "lambda_function" {
+  source  = "terraform-aws-modules/lambda/aws"
+  version = "4.12.1"
+
+  function_name = "${var.namespace}-${var.function_name}"
+  description   = "Persists data from a prepared Grants.gov XML DB extract to DynamoDB."
+
+  role_permissions_boundary         = var.permissions_boundary_arn
+  attach_cloudwatch_logs_policy     = true
+  cloudwatch_logs_retention_in_days = var.log_retention_in_days
+  attach_policy_json                = true
+  policy_json                       = module.lambda_execution_policy.json
+
+  handler       = "bootstrap"
+  runtime       = "provided.al2"
+  architectures = [var.lambda_arch]
+  publish       = true
+  layers        = var.lambda_layer_arns
+
+  source_path = [{
+    path = var.lambda_code_path
+    commands = [
+      "task build-PersistGrantsGovXMLDB",
+      "cd bin/PersistGrantsGovXMLDB",
+      ":zip",
+    ],
+  }]
+  store_on_s3               = true
+  s3_bucket                 = var.lambda_artifact_bucket
+  s3_server_side_encryption = "AES256"
+
+  timeout     = 300 # 5 minutes, in seconds
+  memory_size = 1024
+  environment_variables = merge(var.additional_environment_variables, {
+    DD_TRACE_RATE_LIMIT              = "1000"
+    DD_TAGS                          = join(",", sort([for k, v in local.dd_tags : "${k}:${v}"]))
+    DOWNLOAD_CHUNK_LIMIT             = "20"
+    GRANTS_PREPARED_DATA_BUCKET_NAME = data.aws_s3_bucket.prepared_data.id
+    GRANTS_PREPARED_DYNAMODB_NAME    = var.grants_prepared_dynamodb_table_name
+    LOG_LEVEL                        = var.log_level
+    MAX_CONCURRENT_UPLOADS           = "10"
+    S3_USE_PATH_STYLE                = "true"
+  })
```

Suggest trimming out the unused env vars:
```suggestion
  environment_variables = merge(var.additional_environment_variables, {
    DD_TAGS                          = join(",", sort([for k, v in local.dd_tags : "${k}:${v}"]))
    GRANTS_PREPARED_DYNAMODB_NAME    = var.grants_prepared_dynamodb_table_name
    LOG_LEVEL                        = var.log_level
    S3_USE_PATH_STYLE                = "true"
  })
```

(NB: `DD_TRACE_LIMIT` is probably not necessary for any of these Lambdas; it's set for the `SplitGrantsXMLDB` Lambda function because I'm in the midst of a weeks-long thread with Datadog support surrounding an APM ~bug~ undocumented limitation, and they requested I add it for the time-being).

---

done, this was mostly cruft from when I was retrofitting SplitGrantsXMLDB to fit my needs